src/m_private_utils.h

```diff
@@ -108,9 +108,15 @@
 #endif
 
 
-/* -------------------------MSVC compat defines --------------------- */
-#ifdef _MSC_VER
-# define snprintf _snprintf
+    /* safe cross-platform alternatives to snprintf and vsnprintf.
+    NOTE: if the buffer is too small, both functions return a value >= 'size';
+    however, this is not necessarily the number of required bytes! */
+#ifdef _WIN32
+int pd_snprintf(char *buf, size_t size, const char *format, ...);
+int pd_vsnprintf(char *buf, size_t size, const char *format, va_list argptr);
+#else
+# define pd_snprintf snprintf
+# define pd_vsnprintf vsnprintf
```

Comment:
hmm, i'm not sure whether `m_private_utils.h` is the right place for this (scratch this: i'm sure it is not the right place), if the symbol must be exported.
`m_imp.h` (or the obvious `s_stuff.h`) would be better choices.
also, it should be marked `EXTERN`.

but if we export the symbol, then it is part of the (semi-private) API, and I think the API should be the same on all platforms. so I would do this instead:

```C
EXTERN int pd_snprintf(char *buf, size_t size, const char *format, ...);
EXTERN int pd_vsnprintf(char *buf, size_t size, const char *format, va_list argptr);
#ifndef _WIN32
# define pd_snprintf snprintf
# define pd_vsnprintf vsnprintf
#endif
```

Reply:
> hmm, i'm not sure whether m_private_utils.h is the right place for this (scratch this: i'm sure it is not the right place), if the symbol must be exported.

But the symbol must/should *not* be exported resp. marked `EXTERN`! For now it is meant to be a private helper function. I am not sure about `m_imp.h`, as it seems to be the place for Pd implementation details. Since `pd_snprintf` is a pure utility function, `m_private_utils.h` looks like the right place to me...

Reply:
Ah well, I guess, I see your point. Since I also use it in `pd~`, the symbol needs to be public. After all, we might stop exporting all symbols in the future...

Reply:
yes.

if it wasn't for externals (like `pd~`) then `m_private_utils.h` would of course be fine.

Reply:
I think `s_stuff.h` is the better place.

However, we only really need the symbol for Windows, so the following should be ok:
```
#ifdef _WIN32
EXTERN int pd_snprintf(char *buf, size_t size, const char *format, ...);
EXTERN int pd_vsnprintf(char *buf, size_t size, const char *format, va_list argptr);
#else
# define pd_snprintf snprintf
# define pd_vsnprintf vsnprintf
#endif
```

Reply:
i know that it is only required on Windows, but i don't think this is a reason to have symbols exported on one platform but not on another one.

Reply:
Why not? Why should we export a symbol that we don't even need?

If we export `pd_snprintf` on all platforms, I would rather get rid of the macros alltogether. On advantage would be that non-Windows platforms wouldn't need to include `<stdio.h>`, so the behavior would be more consistent. What do you think?

Reply:
Yes, that's better than having an asymmetric API.

(We could still use a macro in performance-critical code, if we wanted to...)